quinn/src/connection.rs

```diff
@@ -79,29 +79,47 @@ impl Connecting {
 
     /// Convert into a 0-RTT or 0.5-RTT connection at the cost of weakened security
     ///
-    /// Opens up the connection for use before the handshake finishes, allowing the API user to
-    /// send data with 0-RTT encryption if the necessary key material is available. This is useful
-    /// for reducing start-up latency by beginning transmission of application data without waiting
-    /// for the handshake's cryptographic security guarantees to be established.
+    /// Returns `Ok` immediately if the local endpoint is able to attempt 0/0.5-RTT conversion. If
+    /// so, the returned [`Connection`] can be used to open streams and write data to them without
+    /// waiting for the rest of the handshake to complete, at the cost of weakened cryptographic
+    /// security guarantees. The returned [`ZeroRttAccepted`] future resolves when the handshake
+    /// does complete, at which point subsequently opened streams and written data will have full
+    /// cryptographic protection.
     ///
-    /// When the `ZeroRttAccepted` future completes, the connection has been fully established.
+    /// ## Outgoing
     ///
-    /// # Security
+    /// For outgoing connections, the initial attempt to convert to 0-RTT will proceed if the
+    /// [`crypto::ClientConfig`][crate::crypto::ClientConfig] attempts to resume a previous TLS
+    /// session. However, **the remote endpoint may not actually _accept_ the 0-RTT data**--yet
+    /// still accept the connection attempt in general. This possibility is conveyed through the
+    /// [`ZeroRttAccepted`] future--when the handshake completes, it resolves to true if the 0-RTT
+    /// data was accepted and false if it was rejected. If it was rejected, streams opened prior
+    /// to the handshake completing will _hang_, and their existence will typically not be conveyed
+    /// to the remote application.
```

Comment:
Rejected 0-RTT streams should fail immediately with `ZeroRttRejected` errors, not hang. If this isn't happening, please open a bug.

Reply:
> Also, "typically not be conveyed" is confusing, because it suggests that sometimes they will be made visible, which I think is false.

I wrote this as such because that's ultimately an implementation detail of the software on the other side. It hypothetically could have some QUIC implementation that conveys information about rejected 0-RTT data. I could rephrase to make it more clear that that's what I'm talking about, or alternatively, just remove 'typically' altogether if we think it's not a detail worth mentioning?

Reply:
> or alternatively, just remove 'typically' altogether if we think it's not a detail worth mentioning?

Let's. I think it's more confusing than helpful.

Reply:
> Rejected 0-RTT streams should fail immediately with `ZeroRttRejected` errors, not hang. If this isn't happening, please open a bug.

Opened #1904.

Reply:
Looks like a typical missed wakeup bug; opened #1905 to fix.